mitodl/open-discussions
# Make Session auth take priority over JWT tokens

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #1100

#### What's this PR do?

- Swaps the order of the authentication classes
- This caused a change to status code handling for unauthenticated users: status codes that used to be 401 changed to 403 as the first authentication class error handling is used if no auth is provided and a 401 does not make sense for session auth (See https://github.com/encode/django-rest-framework/issues/5968#issuecomment-386519574 and http://www.django-rest-framework.org/api-guide/authentication/#unauthorized-and-forbidden-responses)
- Added an `error_type` field to error responses that serializes the value of `exc.__class__.__name__`
- Updated frontend checks against status codes of 401 and 403 to use this new field (`PostPage.js` and `fetch_auth.js` primarily).

#### How should this be manually tested?
- Verify the app works as usual as a logged in user
  - Particularly, verify that you get the `NotAuthenticated` page if you try to view a private channel you're not a contributor on
- Verify as an unauthenticated user the app works as usual
  - Particularly, verify if you try to visit a channel/post that is private you are prompted to login
